iceoryx_utils/test/moduletests/test_cxx_function_ref.cpp

```diff
@@ -0,0 +1,276 @@
+// Copyright (c) 2020 by Robert Bosch GmbH. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "iceoryx_utils/cxx/function_ref.hpp"
+#include "test.hpp"
+
+using namespace ::testing;
+using namespace iox::cxx;
+
+constexpr int freeFuncTestValue = 42 + 42;
+constexpr int functorTestValue = 11;
+constexpr int memberFuncTestValue = 4273;
+
+int freeFunction()
+{
+    return freeFuncTestValue;
+}
+
+class Functor
+{
+  public:
+    int operator()()
+    {
+        return functorTestValue;
+    }
+};
+
+struct ComplexType
+{
+    char a;
+    int b;
+    float c;
+
+    bool operator==(const ComplexType& rhs) const
+    {
+        if (a == rhs.a && b == rhs.b && c == rhs.c)
+        {
+            return true;
+        }
+        else
+        {
+            return false;
+        }
+    }
+};
+
+ComplexType returnComplexType(ComplexType foo)
+{
+    return foo;
+}
+
+class function_refTest : public Test
+{
+  public:
+    void SetUp() override
+    {
+    }
+
+    void TearDown() override
+    {
+    }
+
+    int foobar()
+    {
+        return memberFuncTestValue;
+    }
+
+    uint8_t m_iter{0};
+};
+
+using function_refDeathTest = function_refTest;
+
+TEST_F(function_refTest, CreateEmptyIsFalse)
+{
+    function_ref<void()> sut;
+    EXPECT_FALSE(sut);
+}
+
+TEST_F(function_refDeathTest, CallEmptyLeadsToTermination)
+{
+    function_ref<void()> sut;
+    EXPECT_DEATH(sut(), ".*");
+}
+
+TEST_F(function_refTest, CreateValidByAssignIsTrue)
+{
+    auto lambda = [] {};
+    function_ref<void()> sut;
+    sut = lambda;
+    EXPECT_TRUE(sut);
+}
+
+TEST_F(function_refTest, CreateValidByAssignResultEqual)
+{
+    auto lambda = []() -> int { return 7253; };
+    function_ref<int()> sut;
+    sut = lambda;
+    EXPECT_THAT(sut(), Eq(7253));
+}
+
+TEST_F(function_refTest, CreateValidByCopyConstructResultEqual)
+{
+    auto lambda = []() -> int { return 3527; };
+    function_ref<int()> sut1{lambda};
+    function_ref<int()> sut2(sut1);
+    EXPECT_TRUE(sut2);
```

Comment:
please make this an ASSERT_TRUE, else the whole test will terminate in the next call

Reply:
Will be done in follow-up #131